src/main/java/org/wso2/carbon/identity/application/authenticator/social/naver/NaverCustomAuthenticator.java

```diff
@@ -0,0 +1,433 @@
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
+import org.wso2.carbon.identity.application.common.model.Property;
+import org.wso2.carbon.identity.core.util.IdentityIOStreamUtils;
+
+public class NaverCustomAuthenticator extends AbstractApplicationAuthenticator
+		implements FederatedApplicationAuthenticator {
+
+	private static final long serialVersionUID = 8654763286341993633L;
+	private static final Log logger = LogFactory.getLog(NaverCustomAuthenticator.class);
+
+	private String stateToken;
+
+	@Override
+	protected void initiateAuthenticationRequest(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationContext context) throws AuthenticationFailedException {
+		logger.info("initiateAuthenticationRequest");
+
+		String stateToken = generateState();
+		this.stateToken = stateToken;
+
+		try {
+			Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+			String clientId = authenticatorProperties.get(NaverCustomAuthenticatorConstants.CLIENT_ID);
+			String authorizationEP = getAuthorizationServerEndpoint();
+			String callbackUrl = authenticatorProperties.get(NaverCustomAuthenticatorConstants.CALLBACK_URL);
+
+			context.setContextIdentifier(stateToken);
+
+			OAuthClientRequest authzRequest = OAuthClientRequest.authorizationLocation(authorizationEP)
+					.setClientId(clientId).setResponseType("code").setRedirectURI(callbackUrl).setState(stateToken)
+					.buildQueryMessage();
+
+			logger.info("authzRequest");
+			logger.info(authzRequest.getLocationUri());
+
+			response.sendRedirect(authzRequest.getLocationUri());
+			logger.info("success");
+		} catch (IOException e) {
+			logger.error("Exception while sending to the login page.", e);
+			throw new AuthenticationFailedException(e.getMessage(), e);
+		} catch (OAuthSystemException e) {
+			logger.error("Exception while building authorization code request.", e);
+			throw new AuthenticationFailedException(e.getMessage(), e);
+		}
+		return;
+	}
+
+	@Override
+	protected void processAuthenticationResponse(HttpServletRequest request, HttpServletResponse response,
+			AuthenticationContext context) throws AuthenticationFailedException {
+
+		logger.info("processAuthenticationResponse");
+		logger.trace("InNaverebookAuthenticator.authenticate()");
+
+		try {
+			if (validateStatusToken(request)) {
+
+				Map<String, String> authenticatorProperties = context.getAuthenticatorProperties();
+				String clientId = authenticatorProperties.get(NaverCustomAuthenticatorConstants.CLIENT_ID);
+				String clientSecret = authenticatorProperties.get(NaverCustomAuthenticatorConstants.CLIENT_SECRET);
+				String tokenEndPoint = getTokenEndpoint();
+
+				String code = getAuthorizationCode(request);
+				String token = getToken(tokenEndPoint, clientId, clientSecret, code);
+				String nvauthUserInfoUrl = getUserInfoEndpoint();
+
+				logger.debug("code >>>>>>> " + code);
+				logger.debug("token >>>>>>> " + token);
```

Comment:
Shall we put this in a audit log, lets make sure whether its safe to log this token and code here